README.md
estimate_genome_size.pl
======

this is a suite of perl scripts to help estimate genome size from KMER distributions. It has been tested with the JELLYFISH program
http://www.cbcb.umd.edu/software/jellyfish/

The output looks something like:

    Estimated Coverage: 59.5 (60X) Estimated Genome Size: 1010144681.14286 (1GB) 

INSTALLATION
------------

To install this script and documentation type the following:

    perl Makefile.PL
    make
    make install

RUN
---

    estimate_genome_size.pl [--version] [--help] --kmer=KMERLEN --peak=PEAK --fastq=fastq [--fastq=fastq]

for detailed documenation

    perldoc estimate_genome_size.pl
    perldoc find_valleys.pl
    perldoc jellyplot.pl

DEPENDENCIES
------------

This module requires Perl

You will also want JELLYFISH:

    http://www.cbcb.umd.edu/software/jellyfish/

For the jellyplot program you will need gnuplot in your path:

    http://www.gnuplot.info/


A TYPICAL SESSION
------------

    # count k-mers (see jellyfish documentation for options)
    gzip -dc reads1.fastq.gz reads2.fastq.gz | jellyfish count -m 31 -o fastq.counts -C -s 10000000000 -U 500 -t 30 /dev/fd/0 
    
    # generate a histogram
    jellyfish histo fastq.counts_0 > fastq.counts_0.histo

    # generate a pdf graph of the histogram
    jellyplot.pl fastq.counts_0.histo

    # look at fastq.counts_0.histo.pdf and identify the approximate peak

    # use find_valleys.pl to help pinpoint the actual peak
    find_valleys.pl fastq.counts_0.histo

    # estimate the size and coverage
    estimate_genome_size.pl --kmer=31 --peak=42 --fastq=reads1.fastq.gz reads2.fastq.gz
    
SEE THE FOLLOWING FOR THE PRINCIPLE BEHIND THE SCRIPT
------------

https://banana-slug.soe.ucsc.edu/bioinformatic_tools:jellyfish
http://seqanswers.com/forums/archive/index.php/t-10988.html
https://banana-slug.soe.ucsc.edu/bioinformatic_tools:quake

IF YOU DON'T BELIEVE IT WORKS
------------

Simulate some next generation data based on an already-sequenced genome.
I used the program ART to do this with the Hydra magnipapillata genome:

http://www.niehs.nih.gov/research/resources/software/biostatistics/art/

    art_illumina --paired --in Hm_genomic.fa --out Hm_genomic_wildtype --len 100 --fcov 43 --mflen 180 --sdev 3.5
    
I also simulated perfect paired-end data with a sliding window. In both cases I recovered the correct coverage and size.

COPYRIGHT AND LICENSE
------------

Copyright (C) 2012,2013 Joseph F. Ryan

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program in the file gpl.txt.  If not, see
http://www.gnu.org/licenses/.



Changes
Revision history for Perl progrom baa.pl.

0.03  Tue Jan 29 09:50:22 CET 2013
	- original public version



jellyplot.pl
#!perl

# THIS PROGRAM REQUIRES GNUPLOT TO BE IN YOUR PATH
# http://www.gnuplot.info/

# Copyright (C) 2012,2013 Joseph F. Ryan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

$|++;

use strict;
use warnings;
use Getopt::Long;
use Pod::Usage;
use IO::File;
use POSIX qw(tmpnam);

our $VERSION = 0.03;
our $AUTHOR  = 'Joseph F. Ryan <jfryan@yahoo.com>';

our $TMP_CMD_FILE = '';
our $TMP_TAB_FILE = '';
our $GNUPLOT = 'gnuplot';

MAIN: {
    my $opt_version = 0;
    my $opt_help = 0;

    my $opt_results = Getopt::Long::GetOptions(  "version" => \$opt_version,
                                                    "help" => \$opt_help);
    $opt_version && version();
    pod2usage({-exitval => 0, -verbose => 2}) if $opt_help;

    my $file = $ARGV[0] or usage();
    make_cmd_file($file);

    if (system("$GNUPLOT $TMP_CMD_FILE\n") == 0) {
        print "pdf file written to $file.pdf\n";
    } else {
        die "system call to $GNUPLOT failed.  Is $GNUPLOT in path?\n";
    }
}

sub make_cmd_file {
    my $file = shift;
    my $fh = '';
    my $date_str = localtime();
    do { $TMP_CMD_FILE = tmpnam() }
    until $fh = IO::File->new($TMP_CMD_FILE, O_RDWR|O_CREAT|O_EXCL);
    print $fh "set terminal postscript\n";
    print $fh "set output \"| ps2pdf - $file.pdf\"\n";
    print $fh "set nokey\n";
    print $fh "set title 'K-mer plot $file (plot generated on $date_str)'\n";
    print $fh "set logscale y\n";
    print $fh "set xlabel 'multiplicity'\n";
    print $fh "set ylabel 'number-of-distinct k-mers with given multiplicity'\n";
    print $fh "plot '$file\n";
}

sub usage {
    die "usage: $0 [--version] [--help] HISTO_FILE\n";
}

sub version {
    die "jellyplot.pl $VERSION\n";
}

# install atexit-style handler so that when we exit or die,
# we automatically delete this temporary file
END { 
    if ($TMP_CMD_FILE) {
        unlink($TMP_CMD_FILE) or die "Couldn't unlink $TMP_CMD_FILE:$!"
    }
    if ($TMP_TAB_FILE) {
        unlink($TMP_TAB_FILE) or die "Couldn't unlink $TMP_TAB_FILE:$!"
    }
}

__END__

=head1 NAME

B<jellyplot.pl> - generate PDF plot of histogram

=head1 AUTHOR

Joseph F. Ryan <josephryan@yahoo.com>

=head1 SYNOPSIS

jellyplot.pl HISTO_FILE

=head1 DESCRIPTION

generate PDF plot of histogram data from jellyfish data. Requires gnuplot

=head1 BUGS

Please report them to <josephryan@yahoo.com>

=head1 COPYRIGHT

Copyright (C) 2012 Joseph F. Ryan 

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


find_valleys.pl
#!perl

# Copyright (C) 2012,2013 Joseph F. Ryan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use Getopt::Long;
use Pod::Usage;

our $VERSION = 0.03;
our $AUTHOR  = 'Joseph F. Ryan <jfryan@yahoo.com>';

MAIN: {
    my $opt_version = 0;
    my $opt_help = 0;
    my $opt_results = Getopt::Long::GetOptions(  "version" => \$opt_version,
                                      "help" => \$opt_help);

    $opt_version && version();
    pod2usage({-exitval => 0, -verbose => 2}) if $opt_help;

    my $file = $ARGV[0] or usage();
    open IN, $file or die "cannot open $file:$!";
    my $first = <IN>;
    chomp $first;
    $first =~ m/^\s*\d+\s+(\d+)\s*$/ || die "unexpected input file format\n";
    my $val = $1;
    my $last_line = $first;
    while (my $line = <IN>) {
        chomp $line;
        $line =~ m/^\s*\d+\s+(\d+)\s*$/ || die "unexpected input file format\n";
        my $current_val = $1;
        if ($current_val > $val) {
            my $diff = $current_val - $val;
            print "valley: $line (diff = $diff) - LAST: $last_line\n";
        }
        $last_line = $line;
        $val = $current_val;
    }
}

sub usage {
    die "usage: $0 [--version] [--help] HISTO_FILE\n";
}

sub version {
    die "find_valleys.pl $VERSION\n";
}

__END__

=head1 NAME

B<find_valleys.pl> - identify possible peaks in histogram data

=head1 AUTHOR

Joseph F. Ryan <josephryan@yahoo.com>

=head1 SYNOPSIS

find_valleys.pl HISTO_FILE

=head1 DESCRIPTION

The script takes histogram file from jellyfish and identifies places where
the values begin to descend. It will print the descending coordinates as well
as the previous coordinates. It is suggested to interpret these results by
examining the actual histogram file and a plot generated by jellyplot.pl

=head1 BUGS

Please report them to <josephryan@yahoo.com>

=head1 COPYRIGHT

Copyright (C) 2012 Joseph F. Ryan 

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut




estimate_genome_size.pl
#!perl

# estimate_genome_size.pl 
#
# estimate coverage and genome size from kmer distribution
# Copyright (C) 2012,2013 Joseph F. Ryan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use Getopt::Long;
use Pod::Usage;

our $VERSION = 0.03;
our $AUTHOR  = 'Joseph F. Ryan <jfryan@yahoo.com>';

MAIN: {
    my $kmer = '';
    my $peak = '';
    my $opt_version = 0;
    my $opt_help    = 0;
    my @fastq = ();

    my $result = GetOptions('kmer=i' => \$kmer,
                           'fastq=s' => \@fastq,
                            'peak=f' => \$peak,
                           'version' => \$opt_version,
                              'help' => \$opt_help);
    $opt_version && version();
    pod2usage({-exitval => 0, -verbose => 2}) if $opt_help;
    die usage() unless ($kmer && $peak && @fastq);
    my ($readlen,$total_nts) = get_readlen_and_total_nts(\@fastq);
    my $coverage = ($peak * $readlen) / ($readlen - $kmer + 1);
    my $genome_size = $total_nts / $coverage;
    print "# running version $VERSION of estimate_genome_size.pl\n";
    print "# run with this command: $0 @ARGV\n";
    print "\n";
    print "TOTAL_NTS: $total_nts\n";
    print "Estimated Coverage: $coverage\n";
    print "Estimated Genome Size: $genome_size\n";
}

sub get_readlen_and_total_nts {
    my $ra_files = shift;
    my $counter = 0;
     
    my $readlen = 0;
    foreach my $file (@{$ra_files}) {
        $counter += 2;
        open IN, $file or die "cannot open $file:$!";
        my $devnull = <IN>;
        my $read = <IN>;
        if ($counter == 1) {
            $readlen = length($read);
        } else {
            die "fastqs have different readlens" unless ($readlen = length($read));
        }
        while (<IN>) {
            $counter ++;
        }
    }
    my $total = ($counter / 4) * $readlen;
    return ($readlen,$total);
}
                            
sub version {
    die "estimate_genome_size.pl $VERSION\n";
}

sub usage {
    die "$0 --kmer=KMERLEN --peak=PEAK --fastq=fastq [--fastq=fastq]\n";
}

__END__

=head1 NAME

B<estimate_genome_size.pl> - use jellyfish peak and fastq files to estimate genome size and coverage

=head1 AUTHOR

Joseph F. Ryan <josephryan@yahoo.com>

=head1 SYNOPSIS

estimate_genome_size.pl --kmer=KMERLEN --peak=PEAK --fastq=FASTQ [--fastq=FASTQ]

=head1 DESCRIPTION

After running jellyfish with a particular KMERLEN and one or more FASTQ files,
determine the PEAK using jellyplot.pl and find_valleys.pl. Next, use this
PEAK as well as the KMERLEN and the FASTQ files used in the jellyfish run
as input. The script will determine the coverage and genome size.

=head1 TYPICAL SESSION

=over 2

# count k-mers (see jellyfish documentation for options)
gzip -dc reads1.fastq.gz reads2.fastq.gz | jellyfish count -m 31 -o fastq.counts -C -s 10000000000 -U 500 -t 30 /dev/fd/0

# generate a histogram
jellyfish histo fastq.counts_0 > fastq.counts_0.histo

# generate a pdf graph of the histogram
jellyplot.pl fastq.counts_0.histo

# look at fastq.counts_0.histo.pdf and identify the approximate peak

# use find_valleys.pl to help pinpoint the actual peak
find_valleys.pl fastq.counts_0.histo

# estimate the size and coverage
estimate_genome_size.pl --kmer=31 --peak=42 --fastq=reads1.fastq.gz reads2.fastq.gz

=back

=head1 BUGS

Please report them to <josephryan@yahoo.com>

=head1 COPYRIGHT

Copyright (C) 2012 Joseph F. Ryan 

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut

